src/cargo/ops/cargo_compile.rs

```diff
@@ -404,12 +404,7 @@ pub fn create_bcx<'a, 'cfg>(
                 .shell()
                 .warn("-Zbuild-std does not currently fully support --build-plan")?;
         }
-        if build_config.requested_kinds[0].is_host() {
-            // TODO: This should eventually be fixed. Unfortunately it is not
-            // easy to get the host triple in BuildConfig. Consider changing
-            // requested_target to an enum, or some other approach.
-            anyhow::bail!("-Zbuild-std requires --target");
```

Comment:
I think that we still need to keep this around in one form or another. IIRC the issue has to do with where if you have `cargo build` for nothing with a per-package-target or anything like that then build scripts and proc macros will do the wrong thing and try to use the custom-built libstd. 

I think though that this check would need to improve because if your manifest has a per-package-target then there's no need to require `--target` on the command line, so this check may just need to get deferred to later?

Reply:
The todo comment seems to suggest that we would like to see support for `-Zbuild-std`ing the host target though, which would be supported after this change thanks to target selection overhaul in our logic.

> then build scripts and proc macros will do the wrong thing and try to use the custom-built libstd.

Ah I see. Then we could avoid attaching the custom std dependencies for proc-macro and build script runs. I will write a test for this.

Reply:
On a second glance I don't think we should add this as a test. We can't just do `-Zbuild-std=std` because then it doesn't matter if a dependent crate is proc-macro of a build script because what happens if they use the custom std?

But if we do `-Zbuild-std=core`, the dependent crate will use the default sysroot instead, IIUC. (because it wants to use proc-macro and `build-std` is not configured to build `proc-macro` therefore it uses the installed libstd). On my local machine I did try to add some tests but it fails when I add them:

```
error: no matching package named `pm` found
location searched: registry `crates-io`
required by package `foo v0.1.0 (/shared/develop/cargo/target/tmp/cit/t0/foo)`
', tests/build-std/main.rs:353:70
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    proc_macro_dep_no_explicit_target
```

Reply:
There were a number of blockers to enabling `cargo build -Zbuild-std`. This PR is one of them but it is not the only blocker, and the other blocker (that I know of at least) is that host things (build scripts and proc macros) use the custom-built std. Build scripts and proc macros should always use the host-defined std, never the custom-built std. One major use case is rebuilding the custom-built std with different codegen options that may or may not be applicable to the host (e.g. fuzz options). 

I realize that requiring `--target` obviates this PR somewhat, but this is not a conditional that can be unconditionally removed until this other issue in the behavior is fixed.

Reply:
I'm stuck. I think we could continue the last PR but keep it also as a cli option? Then we could rewrite the logic handling each crate instead of the current way, handling the whole build.

Reply:
To clarify, I think "fixing" this to not actually need `--target` is likely to be somewhat difficult in Cargo. I haven't looked myself to see what it might entail but I don't think it will be trivial.

Otherwise I don't have a preference to go about this. My main point is that we can't simply stop requiring `--target` due to the consequence of how Cargo treats things like `RUSTFLAGS`